ethpm/tools/builder.py

```diff
@@ -298,39 +291,109 @@ def _contract_type(
         raise ManifestBuildingError(
             "Duplicate contract names were found in the compiler output."
         )
     return {
-        name: normalize_contract_type(compiler_output[":".join((path, name))])
+        name: normalize_contract_type(compiler_output[path][name])
         for path, name in paths_and_names
     }
 
 
 @to_dict
 def normalize_contract_type(
     contract_type_data: Dict[str, Any]
-) -> Generator[Dict[str, Any], None, None]:
+) -> Iterable[Tuple[str, Any]]:
     """
     Serialize contract_data found in compiler output to the defined fields.
     """
-    yield "abi", json.loads(contract_type_data["abi"])
-    yield "deployment_bytecode", {"bytecode": add_0x_prefix(contract_type_data["bin"])}
-    yield "natspec", json.loads(contract_type_data["devdoc"])
+    yield "abi", contract_type_data["abi"]
+    yield "deployment_bytecode", normalize_bytecode_object(
+        contract_type_data["evm"]["bytecode"]
+    )
+    # todo support natspec, runtime_bytecode
+    yield "natspec", {}
+
+
+@to_dict
+def normalize_bytecode_object(obj: Dict[str, Any]) -> Iterable[Tuple[str, Any]]:
+    if obj["linkReferences"]:
+        yield "link_references", process_link_references(
+            obj["linkReferences"], obj["object"]
+        )
+        yield "bytecode", process_bytecode(obj["linkReferences"], obj["object"])
+    else:
+        yield "bytecode", add_0x_prefix(obj["object"])
+
+
+def process_bytecode(link_refs: Dict[str, Any], bytecode: bytes) -> bytes:
+    """
+    Replace link_refs in bytecode with 0's.
+    """
+    all_offsets = [y for x in link_refs.values() for y in x.values()]
+    link_fns = (
+        replace_link_ref_in_bytecode(ref["start"] * 2, ref["length"] * 2)
+        for ref in cytoolz.concat(all_offsets)
+    )
+    processed_bytecode = cytoolz.pipe(bytecode, *link_fns)
+    return add_0x_prefix(processed_bytecode)
+
+
+@cytoolz.curry
+def replace_link_ref_in_bytecode(offset: int, length: int, bytecode: str) -> str:
+    validate_link_ref(offset, length, bytecode)
+    new_bytes = (
+        bytecode[:offset] + "0" * length + bytecode[offset + length :]  # noqa: E203
+    )
+    return new_bytes
+
+
+# todo pull all bytecode linking/validating across py-ethpm into shared utils
+@to_list
+def process_link_references(
+    link_refs: Dict[str, Any], bytecode: str
+) -> Iterable[List[Dict[str, Any]]]:
+    for link_ref in link_refs.values():
+        yield normalize_link_ref(link_ref, bytecode)
+
+
+def normalize_link_ref(link_ref: Dict[str, Any], bytecode: str) -> Dict[str, Any]:
+    name = list(link_ref.keys())[0]
+    return {
+        "name": name,
+        "length": 20,
+        "offsets": normalize_offsets(link_ref, bytecode),
+    }
+
+
+@to_list
+def normalize_offsets(data: Dict[str, Any], bytecode: str) -> Iterable[List[int]]:
+    for link_ref in data.values():
+        for ref in link_ref:
+            validate_link_ref(ref["start"] * 2, ref["length"] * 2, bytecode)
```

Comment:
I still think it's an anti-pattern to mix validation and normalization.  There's nothing inherently bad about this but I *think* it would be cleaner to keep these separate.

1. generate the normalized link_ref objects.
2. validate the offsets (all at once) against the original bytecode.

Reply:
Gotcha, moved the link_ref validation to it's own process on lines 347-351